File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def checkEqual(root1, root2):
            if root1.left and root2.left:
                if checkEqual(root1.left, root2.left) is False:
                    return False
            elif root1.left or root2.left:
                return False

            if root1.right and root2.right:
                if checkEqual(root1.right, root2.right) is False:
                    return False
            elif root1.right or root2.right:
                return False

            if root1.val != root2.val:
                return False

        def dfs(root):
            if root.left:
                if dfs(root.left):
                    return True
            if root.right:
                if dfs(root.right):
                    return True

            if root.val == subRoot.val:
               if checkEqual(root, subRoot) != False:
                # print("return true")
                return True

            print("val -> ", root.val)
        if root:
            ret = dfs(root)
            if ret is None:
                return False
            return ret

        return False

File: Data Structures & Algorithms/depth-of-binary-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 * int val;
 * TreeNode *left;
 * TreeNode *right;
 * TreeNode() : val(0), left(nullptr), right(nullptr) {}
 * TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int max_depth = 0;

    int maxDepth(TreeNode* root) {
        max_depth = 0; // Reset for each call
        if (root != nullptr) {
            goDepth(1, root);
        }
        return max_depth;
    }

    void goDepth(int n, TreeNode* root) {
        // Update the global maximum depth
        if (n > max_depth) {
            max_depth = n;
        }

        // Recursive calls for children
        if (root->left != nullptr) {
            goDepth(n + 1, root->left);
        }
        if (root->right != nullptr) {
            goDepth(n + 1, root->right);
        }
    }
};